### PR TITLE
boot_serial: Fix incorrect format specifier for off_t

### PR DESCRIPTION
**Pull request by stephanosio**
The `BOOT_LOG_INF` function, which uses the format specifiers defined
by the C standard, was incorrectly printing a variable with the type of
`off_t` using the `%x` format specifier, which is intended to be used
with the `int` type.

The `off_t` type, specified by the POSIX standard, is not guaranteed to
be `int`, and it may be defined as `long` or `long long` depending on
the toolchain and the target architecture.

This commit updates the print routine such that it casts the arguments
of the `off_t` type to `intmax_t` and prints them out using the
corresponding `%jx` format specifier.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

### CODE REVIEW

**Comment by stephanosio** [comment]
FYI @nvlsianpu

**Review by gmarull — APPROVED** [review]

**Comment by ngboonkhai** [comment]
Hi, may I know how's the progress of this pull request? Is it pending for another reviewer before it can be merged?

**Comment by ngboonkhai** [comment]
@de-nordic

**Review by d3zd3z — COMMENTED** [review]
I'm a little concerned if the 'j' is available in all of the printf variants that might be available to us. It doesn't look like nanolibc supports it, and I wonder if this will work in those configurations.

Intmax might also be fairly inefficient on some of these targets, but I'm not sure of a better way to handle it, at least portably.

File: boot/boot_serial/src/boot_serial.c
Comment (de-nordic) [inline boot/boot_serial/src/boot_serial.c:304]: Why not to cast this to `ssize` and use `z`?

Comment (stephanosio) [reply boot/boot_serial/src/boot_serial.c:304]: because `ssize_t` is not guaranteed to be the largest integer type supported and/or the same size as `off_t`.

**Comment by stephanosio** [comment]
> I'm a little concerned if the 'j' is available in all of the printf variants that might be available to us.

The `j` modifier was introduced in the C99 and it **should be supported** by all toolchains/libc that claim to support the C99 standard and above.

Also, as far as Zephyr is concerned, it is supported by all toolchains/libc; but, as you noted, that might not be the case for the rest of the targets that mcuboot supports.

If there indeed is a toolchain/libc supported by the mcuboot that does not support the `j` modifier, then I will revise this patch to cast the values to `(long)` and do `%lx`, which is far from ideal but should be portable ...

File: boot/boot_serial/src/boot_serial.c
Comment (de-nordic) [reply boot/boot_serial/src/boot_serial.c:304]: > because `ssize_t` is not guaranteed to be the largest integer type supported and/or the same size as `off_t`.

We should probably replace the `off_t`, here,  with the `uint32_t`, including the `erase_range` definition; I do not think that we need offsets and sizes greater than 32-bit here (for example Zephyr flash_map sizes are `uint32_t`). Than just use `llx` and cast expressions to `unsigned long long` - this should cover both 32 and 64 bit systems for us.

**Comment by stephanosio** [comment]
> If there indeed is a toolchain/libc supported by the mcuboot that does not support the `j` modifier

That is "if there indeed is a toolchain/libc supported by the mcuboot that does not support the j modifier."

Someone will need to approve the CI run here so we can see.

**Comment by ngboonkhai** [comment]
@nvlsianpu @de-nordic

**Comment by ngboonkhai** [comment]
@nvlsianpu @de-nordic @ioannisg @utzig 
Please help to approve the CI run and get this resolved asap, this PR is blocking 
https://github.com/zephyrproject-rtos/zephyr/pull/47612
https://github.com/zephyrproject-rtos/zephyr/pull/46443

**Comment by ngboonkhai** [comment]
> > @nvlsianpu @de-nordic @ioannisg @utzig @d3zd3z Please help to approve the CI run and get this resolved asap, this PR is blocking [zephyrproject-rtos/zephyr#47612](https://github.com/zephyrproject-rtos/zephyr/pull/47612) [zephyrproject-rtos/zephyr#46443](https://github.com/zephyrproject-rtos/zephyr/pull/46443)
> 
> I do not see how the [zephyrproject-rtos/zephyr#46443](https://github.com/zephyrproject-rtos/zephyr/pull/46443)]([zephyrproject-rtos/zephyr#46443](https://github.com/zephyrproject-rtos/zephyr/pull/46443) is depending on this. Anyway, making change here will not immediately fix zephyrproject-rtos/mcuboot, there needs to be sync done.

Because I did a workaround to just pass the workflow check for the flash_shell.c, so the reviewer can continue review on the other part, once this PR 1413 is checked in will remove the work around patch
https://github.com/zephyrproject-rtos/zephyr/pull/46443/commits/62e1f639ea2552f9f81b53de2a73fe94868cc318

**Comment by ngboonkhai** [comment]
> > > @nvlsianpu @de-nordic @ioannisg @utzig @d3zd3z Please help to approve the CI run and get this resolved asap, this PR is blocking [zephyrproject-rtos/zephyr#47612](https://github.com/zephyrproject-rtos/zephyr/pull/47612) [zephyrproject-rtos/zephyr#46443](https://github.com/zephyrproject-rtos/zephyr/pull/46443)
> > 
> > 
> > I do not see how the [zephyrproject-rtos/zephyr#46443](https://github.com/zephyrproject-rtos/zephyr/pull/46443)]([zephyrproject-rtos/zephyr#46443](https://github.com/zephyrproject-rtos/zephyr/pull/46443) is depending on this. Anyway, making change here will not immediately fix zephyrproject-rtos/mcuboot, there needs to be sync done.
> 
> Because I did a workaround to just pass the workflow check for the flash_shell.c, so the reviewer can continue review on the other part, once this PR 1413 is checked in will remove the work around patch [zephyrproject-rtos/zephyr@62e1f63](https://github.com/zephyrproject-rtos/zephyr/commit/62e1f639ea2552f9f81b53de2a73fe94868cc318)

> Anyway, making change here will not immediately fix zephyrproject-rtos/mcuboot, there needs to be sync done.
That's the reason why I keep pushing.

**Comment by ngboonkhai** [comment]
Hi @de-nordic,

Is @nvlsianpu the maintainer? He seems like not very responsive.

**Comment by carlescufi** [comment]
> Hi @de-nordic,
> 
> Is @nvlsianpu the maintainer? He seems like not very responsive.

He's on holiday.

**Comment by ngboonkhai** [comment]
> > Hi @de-nordic,
> > Is @nvlsianpu the maintainer? He seems like not very responsive.
> 
> He's on holiday.

I see, thanks for informing. Other than him, is there any other colleague that would be able to help?

**Comment by ngboonkhai** [comment]
Hi, @utzig @d3zd3z,

I notice that you are the top contributor of mcuboot, can you guys help to enable the CI workflow to run? Thanks

**Review by gustavonihei — APPROVED** [review]
LGTM

**Comment by ngboonkhai** [comment]
Hi @stephanosio, can you help to take a look at the CI result?

**Comment by ngboonkhai** [comment]
Hi @stephanosio, any updates?

**Comment by stephanosio** [comment]
> Hi @stephanosio, any updates?

This is not a problem specific to this PR. The mcuboot maintainers will need to fix their CI.

**Comment by ngboonkhai** [comment]
Hi, @utzig @d3zd3z @gustavonihei @nvlsianpu @de-nordic @ioannisg. Can you please take a look at the CI?

**Comment by nvlsianpu** [comment]
@ngboonkhai @stephanosio Can you rebase? This looks like the "too big flash footprint" issue we recently fixed on the mian branch.

**Comment by ngboonkhai** [comment]
> @ngboonkhai @stephanosio Can you rebase? This looks like the "too big flash footprint" issue we recently fixed on the mian branch.

Hi @nvlsianpu, Need stephan help to rebase, I'm here to keep track this issue to unblock my PR in Zephyr https://github.com/zephyrproject-rtos/zephyr/pull/46443

**Comment by stephanosio** [comment]
Rebased

**Review by nvlsianpu — APPROVED** [review]

**Comment by ngboonkhai** [comment]
Thanks everyone for he effort. 
Hi @nvlsianpu would you create a PR in zephyr repo to handle the mcuboot fork sync and west.yml ?

**Comment by nvlsianpu** [comment]
@ngboonkhai Yes I will do this on Monday.

**Comment by ngboonkhai** [comment]
> @ngboonkhai Yes I will do this on Monday.

@nvlsianpu  Okay, Thanks a lot :D

**Comment by ngboonkhai** [comment]
> [nvlsianpu](/nvlsianpu)

Hi @nvlsianpu

Quick check with you on the mcuboot fork sync, have you got a chance to look into it? I try my luck to search over the zephyr PR but can't find.

**Comment by nvlsianpu** [comment]
https://github.com/zephyrproject-rtos/mcuboot/pull/85

**Comment by ngboonkhai** [comment]
> [zephyrproject-rtos#85](https://github.com/zephyrproject-rtos/mcuboot/pull/85)

Thanks @nvlsianpu

**Comment by ngboonkhai** [comment]
> [nvlsianpu](/nvlsianpu)

Hi @nvlsianpu I believe you need to update the revision number at the zephyr west.yaml as well
https://github.com/zephyrproject-rtos/zephyr/blob/main/west.yml#L175

**Comment by carlescufi** [comment]
> > [nvlsianpu](/nvlsianpu)
> 
> Hi @nvlsianpu I believe you need to update the revision number at the zephyr west.yaml as well https://github.com/zephyrproject-rtos/zephyr/blob/main/west.yml#L175

https://github.com/zephyrproject-rtos/zephyr/pull/48786

**Pull request merged**
